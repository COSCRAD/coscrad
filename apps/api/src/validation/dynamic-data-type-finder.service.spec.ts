import {
    getCoscradDataSchema,
    NestedDataType,
    NonEmptyString,
    NonNegativeFiniteNumber,
    Union2,
    Union2Member,
} from '@coscrad/data-types';
import { Test } from '@nestjs/testing';
import { DynamicDataTypeFinderService } from './dynamic-data-type-finder.service';
import { DynamicDataTypeModule } from './dynamic-data-type.module';

describe('DynamicDataTypeFinderService', () => {
    let dynamicDataTypeFinderService: DynamicDataTypeFinderService;

    const DUMMY_CONTEXT_UNION = 'DUMMY_CONTEXT_UNION';

    @Union2Member(DUMMY_CONTEXT_UNION, 'one')
    class Context1 {
        type: 'one';

        @NonEmptyString({
            label: 'foo',
            description: 'foo',
        })
        foo: string;
    }

    @Union2Member(DUMMY_CONTEXT_UNION, 'two')
    class Context2 {
        type: 'two';

        @NonNegativeFiniteNumber({
            label: 'bar',
            description: 'bar',
        })
        bar: number;
    }

    class EdgeConnectionMember {
        @NonEmptyString({
            label: 'ID',
            description: 'unique identifier',
        })
        id: string;

        @Union2(DUMMY_CONTEXT_UNION, 'type', {
            label: 'context',
            description: 'contextualizes the connection within the resource context',
        })
        context: Context1 | Context2;
    }

    class EdgeConnection {
        @NestedDataType(EdgeConnectionMember, {
            isArray: true,
            label: `members`,
            description: `information about resources that participate in this connection`,
        })
        members: EdgeConnectionMember[];
    }

    beforeAll(async () => {
        const testModule = await Test.createTestingModule({
            imports: [DynamicDataTypeModule],
            providers: [EdgeConnection, EdgeConnectionMember, Context1, Context2],
        }).compile();

        dynamicDataTypeFinderService = testModule.get<DynamicDataTypeFinderService>(
            DynamicDataTypeFinderService
        );
    });

    describe(`when there is a valid union type definition for a class`, () => {
        it(`should bootstrap the schema definitions`, async () => {
            await dynamicDataTypeFinderService.bootstrapDynamicTypes();

            const dataSchema = getCoscradDataSchema(EdgeConnection);

            const nestedUnionMemberSchemaDefinitions =
                dataSchema['members']['schema']['context']['schemaDefinitions'];

            expect(nestedUnionMemberSchemaDefinitions).toHaveLength(2);

            expect(dataSchema).toMatchSnapshot();
        });
    });
});
