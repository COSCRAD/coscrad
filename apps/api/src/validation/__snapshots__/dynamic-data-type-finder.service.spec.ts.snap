// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicDataTypeFinderService when there is a valid union type definition for a class should bootstrap the schema definitions 1`] = `
{
  "members": {
    "complexDataType": "NESTED_TYPE",
    "description": "information about resources that participate in this connection",
    "isArray": true,
    "isOptional": false,
    "label": "members",
    "schema": {
      "context": {
        "complexDataType": "UNION_TYPE",
        "description": "contextualizes the connection within the resource context",
        "discriminantPath": "type",
        "isArray": false,
        "isOptional": false,
        "label": "context",
        "schemaDefinitions": [
          {
            "discriminantValue": "one",
            "schema": {
              "foo": {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "foo",
                "isArray": false,
                "isOptional": false,
                "label": "foo",
              },
            },
          },
          {
            "discriminantValue": "two",
            "schema": {
              "bar": {
                "coscradDataType": "NON_NEGATIVE_FINITE_NUMBER",
                "description": "bar",
                "isArray": false,
                "isOptional": false,
                "label": "bar",
              },
            },
          },
        ],
        "unionName": "DUMMY_CONTEXT_UNION",
      },
      "id": {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "unique identifier",
        "isArray": false,
        "isOptional": false,
        "label": "ID",
      },
    },
  },
}
`;
